dpnp/dpnp_algo/dpnp_algo_mathematical.pyx

```diff
@@ -273,23 +273,37 @@ cpdef tuple dpnp_modf(dparray x1):
     return result1, result2
 
 
-cpdef dparray dpnp_multiply(dparray x1, x2):
-    cdef dparray result
-    if dpnp.isscalar(x2):
-        x2_ = dpnp.array([x2])
+cpdef dparray dpnp_multiply(dpnp_input1 x1, dpnp_input2 x2):
+    if dpnp_input1 in dpnp_numeric and dpnp_input2 is dparray:
+        return dpnp_multiply_array_scalar(x2, x1)
+    elif dpnp_input1 is dparray and dpnp_input2 in dpnp_numeric:
+        return dpnp_multiply_array_scalar(x1, x2)
+    elif dpnp_input1 is dparray and dpnp_input2 is dparray:
+        return call_fptr_2in_1out(DPNP_FN_MULTIPLY_ARRAY_ARRAY, x1, x2, x1.shape)
+    else:
+        return x1 * x2
```

Comment:
It looks like this code must be inside DPNPC (not at Cython leyer)

Reply:
We have different return type for input `array*array` and `array*scalar` of the same type.